File: client/src/services/calendar.ts
import { API_URL } from '../config'
import axios from 'axios'
import {
  Country,
  SearchParams,
  TournamentCategory,
  TournamentGrade,
  TournamentOrganization,
  TournamentPreview,
  TournamentSearchResponse,
} from '../types/tournament'

const URL = `${API_URL}/tournaments`

const getTournaments = async (): Promise<TournamentPreview[]> => {
  const response = await axios.get(URL)
  return response.data.results
}

const searchTournaments = async (
  search: SearchParams
): Promise<TournamentSearchResponse> => {
  let url = `${URL}/search?`

  if (search.startDate) url += `start=${search.startDate}&`
  if (search.endDate) url += `end=${search.endDate}&`
  if (search.searchText) url += `search=${search.searchText}&`
  if (search.country) url += `country=${search.country}&`
  if (search.categories) url += `categories=${search.categories}&`
  if (search.organization) url += `organization=${search.organization}&`
  if (search.perPage) url += `per_page=${search.perPage}&`
  if (search.page) url += `page=${search.page}`

  const response = await axios.get(url)
  return response.data
}

const getCategories = async (): Promise<TournamentCategory[]> => {
  const response = await axios.get(`${URL}/categories`)
  return response.data.results
}

const getCategoriesByGrade = async (): Promise<TournamentGrade[]> => {
  const response = await axios.get(`${URL}/grades`)
  return response.data.results
}

const getOrganizations = async (): Promise<TournamentOrganization[]> => {
  const response = await axios.get(`${URL}/organizations`)
  return response.data.results
}

const getCountries = async (): Promise<Country[]> => {
  const response = await axios.get(`${URL}/countries`)
  return response.data.results
}

const service = {
  getTournaments,
  searchTournaments,
  getCategories,
  getCategoriesByGrade,
  getOrganizations,
  getCountries,
}

export default service

File: server/src/routes/calendar.py
from flask import Blueprint, jsonify, request
from services import calendar

calendar_bp = Blueprint("tournaments", __name__)


@calendar_bp.route("/", methods=["GET"])
def get_all():
    data = calendar.search()
    return jsonify(data)


@calendar_bp.route("/grades", methods=["GET"])
def get_categories_by_grade():
    # Dates are optional
    start_date = request.args.get("start")
    end_date = request.args.get("end")

    data = calendar.get_categories_by_grade(start_date, end_date)
    return jsonify(data)


@calendar_bp.route("/categories", methods=["GET"])
def get_categories():
    # Dates are optional
    start_date = request.args.get("start")
    end_date = request.args.get("end")

    data = calendar.get_categories(start_date, end_date)
    return jsonify(data)


@calendar_bp.route("/organizations", methods=["GET"])
def get_organizations():
    data = calendar.get_organizations()
    return jsonify(data)


@calendar_bp.route("/countries", methods=["GET"])
def get_countries():
    data = calendar.get_countries()
    return jsonify(data)


@calendar_bp.route("/search", methods=["GET"])
def search():
    categories = request.args.get("categories")
    payload = request.args.to_dict()
    payload["categories"] = categories.split(",") if categories else []

    data = calendar.search(**payload)
    return jsonify(data)
